docs/proposals/20210628-workloadspread.md

```diff
@@ -0,0 +1,287 @@
+last-updated: 2021-06-28
+status: implementable
+---
+
+# WorkloadSpread
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+## Motivation
+
+Define a CRD named WorkloadSpread, which can support configuration and management for elastic and multi-domain applications,
+even more powerful and flexible than [UnitedDeployment](https://openkruise.io/en-us/docs/uniteddeployment.html) in some ways.
+
+Not like the workload type, WorkloadSpread can be used with those stateless workloads, such as CloneSet, Deployment and ReplicaSet.
+
+So users needn't learn the usage of a new workload and how to deploy and update their application by it.
+They only have to create WorkloadSpread CRs related to their existing workloads. Everything is just great!
+
+## Proposal
+
+WorkloadSpread definition looks like this:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: WorkloadSpread
+metadata:
+  namespace: ...
+  name: ...
+  # ...
+spec:
+  targetRef:
+    apiVersion: apps/v1 | apps.kruise.io/v1alpha1
+    kind: Deployment | CloneSet
+    name: workload-xxx
+  subsets:
+  - name: subset-a
+    requiredNodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-a
+    preferredNodeSelectorTerms:
+    - weight: 50
+      preference:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-b
+    maxReplicas: 3
+    tolertions: []
+    patch:
+      metadata:
+        labels:
+          xxx-specific-label: xxx
+  - name: subset-b
+    requiredNodeSelectorTerm:
+      matchExpressions:
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values:
+        - zone-b
+  scheduleStrategy:
+    type: Adaptive | Fixed
+    adaptive:
+      disableSimulationSchedule: false
+      rescheduleCriticalSeconds: 30
+status:
+  observedGeneration: 12
+  subsetStatuses:
+  - name: subset-a
+    missingReplicas: 2
+    creatingPods:
+      pod-xxx: "2021-04-27T08:34:11Z"
+    deletingPods:
+      pod-yyy: "2021-04-27T08:34:11Z"
+    subsetUnscheduledStatus:
+      unschedulable: true
+      unscheduledTime: 2021-04-27T08:34:11Z
+      failedCount: 1
+  - name: subset-b
+    missingReplicas: -1
+```
+
+**spec**:
+- `targetRef`: reference to a specific workload, it supports `Deployment` in `apps` group and `CloneSet` in `apps.kruise.io` group
+- `subsets`: list of subsets which defines different topologies
+  - `name`: name of this subset
+  - `requiredNodeSelectorTerm`: the additional node selector of this subset, it will be injected into each term of `pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution.nodeSelectorTerms`
```

Comment:
> it will be injected into each term of

the nodes corresponding to each term in requiredDuringSchedulingIgnoredDuringExecution are intersected, so is it more appropriate to just append the term into the  requiredNodeSelectorTerm?

Reply:
`pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution.nodeSelectorTerms ` is a list of node selector terms and the terms are ORed. 
So we need to inject `requiredNodeSelectorTerm` into each terms to enforce scheduling Pod to subset.

It is more appropriate to append the term into the `preferredNodeSelectorTerms` .